File: src/main/java/AoC2015_05.java
import java.util.List;
import java.util.regex.Pattern;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2015_05
            extends SolutionBase<List<String>, Integer, Integer> {
    
    private static final Pattern VOWEL = Pattern.compile("(a|e|i|o|u)");
    private static final Pattern TWIN = Pattern.compile("([a-z])\\1");
    private static final Pattern BAD = Pattern.compile("(ab|cd|pq|xy)");
    private static final Pattern TWO_TWINS = Pattern.compile("([a-z]{2})[a-z]*\\1");
    private static final Pattern THREE_LETTER_PALINDROME = Pattern.compile("([a-z])[a-z]\\1");

    private AoC2015_05(final boolean debug) {
        super(debug);
    }

    public static AoC2015_05 create() {
        return new AoC2015_05(false);
    }

    public static AoC2015_05 createDebug() {
        return new AoC2015_05(true);
    }
    
    private int countMatches(final Pattern pattern, final String str) {
        return (int) pattern.matcher(str).results().count();
    }
    
    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }

    @Override
    public Integer solvePart1(final List<String> input) {
        return (int) input.stream()
                .filter(s -> countMatches(VOWEL, s) >= 3)
                .filter(s -> countMatches(TWIN, s) >= 1)
                .filter(s -> countMatches(BAD, s) == 0)
                .count();
    }

    @Override
    public Integer solvePart2(final List<String> input) {
        return (int) input.stream()
                .filter(s -> countMatches(TWO_TWINS, s) >= 1)
                .filter(s -> countMatches(THREE_LETTER_PALINDROME, s) >= 1)
                .count();
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "1"),
        @Sample(method = "part1", input = TEST2, expected = "1"),
        @Sample(method = "part1", input = TEST3, expected = "0"),
        @Sample(method = "part1", input = TEST4, expected = "0"),
        @Sample(method = "part1", input = TEST5, expected = "0"),
        @Sample(method = "part2", input = TEST6, expected = "1"),
        @Sample(method = "part2", input = TEST7, expected = "1"),
        @Sample(method = "part2", input = TEST8, expected = "0"),
        @Sample(method = "part2", input = TEST9, expected = "0"),
        @Sample(method = "part2", input = TEST10, expected = "1"),
    })
    public void samples() {
    }

    public static void main(final String[] args) throws Exception {
        AoC2015_05.create().run();
    }

    private static final String TEST1 = "ugknbfddgicrmopn";
    private static final String TEST2 = "aaa";
    private static final String TEST3 = "jchzalrnumimnmhp";
    private static final String TEST4 = "haegwjzuvuyypxyu";
    private static final String TEST5 = "dvszwmarrgswjxmb";
    private static final String TEST6 = "qjhvhtzxzqqjkmpb";
    private static final String TEST7 = "xxyxx";
    private static final String TEST8 = "uurcxstgmygtbstg";
    private static final String TEST9 = "ieodomkazucvgmuy";
    private static final String TEST10 = "xyxy";

}

File: src/main/cpp/2015/05/AoC2015_05.cpp
#include <assert.h>

#include <regex>
#include <string>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

const string VOWEL = "(a|e|i|o|u)";
const string TWIN = "([a-z])\\1";
const string BAD_PAIR = "(ab|cd|pq|xy)";
const string TWO_TWINS = "([a-z]{2})[a-z]*\\1";
const string THREE_LETTER_PALINDROME = "([a-z])[a-z]\\1";

int countMatches(const string& pattern, const string& s) {
    const regex re(pattern);
    auto begin = sregex_iterator(s.begin(), s.end(), re);
    auto end = sregex_iterator();
    return distance(begin, end);
}

int part1(const vector<string>& input) {
    return count_if(input.begin(), input.end(), [](const string& line) {
        return countMatches(VOWEL, line) >= 3 &&
               countMatches(TWIN, line) >= 1 &&
               countMatches(BAD_PAIR, line) == 0;
    });
}

int part2(const vector<string>& input) {
    return count_if(input.begin(), input.end(), [](const string& line) {
        return countMatches(TWO_TWINS, line) >= 1 &&
               countMatches(THREE_LETTER_PALINDROME, line) >= 1;
    });
}

const vector<string> TEST1 = {"ugknbfddgicrmopn"};
const vector<string> TEST2 = {"aaa"};
const vector<string> TEST3 = {"jchzalrnumimnmhp"};
const vector<string> TEST4 = {"haegwjzuvuyypxyu"};
const vector<string> TEST5 = {"dvszwmarrgswjxmb"};
const vector<string> TEST6 = {"qjhvhtzxzqqjkmpb"};
const vector<string> TEST7 = {"xxyxx"};
const vector<string> TEST8 = {"uurcxstgmygtbstg"};
const vector<string> TEST9 = {"ieodomkazucvgmuy"};
const vector<string> TEST10 = {"xyxy"};

void samples() {
    assert(part1(TEST1) == 1);
    assert(part1(TEST2) == 1);
    assert(part1(TEST3) == 0);
    assert(part1(TEST4) == 0);
    assert(part1(TEST5) == 0);
    assert(part2(TEST6) == 1);
    assert(part2(TEST7) == 1);
    assert(part2(TEST8) == 0);
    assert(part2(TEST9) == 0);
    assert(part2(TEST10) == 1);
}

MAIN(2015, 5)
